python/lsst/pipe/base/butlerQuantumContext.py

```diff
@@ -0,0 +1,204 @@
+        Butler object from/to which datasets will be get/put
+    quantum : `lsst.daf.butler.core.Quantum`
+        Quantum object that describes the data-sets which will
+        be get/put by a single execution of this node in the
+        pipeline graph.
+    """
+    def __init__(self, butler: Butler, quantum: Quantum):
+        self.quantum = quantum
+        self.registry = butler.registry
+        self.allInputs = set()
+        self.allOutputs = set()
+        for refs in quantum.predictedInputs.values():
+            for ref in refs:
+                self.allInputs.add((ref.datasetType, ref.dataId))
+        for refs in quantum.outputs.values():
+            for ref in refs:
+                self.allOutputs.add((ref.datasetType, ref.dataId))
+
+        # Create closures over butler to discourage anyone from directly
+        # using a butler reference
+        def _get(self, ref):
+            if isinstance(ref, DeferredDatasetRef):
+                self._checkMembership(ref.datasetRef, RefDirection.IN)
+                return butler.getDeferred(ref.datasetRef)
+
+            else:
+                self._checkMembership(ref, RefDirection.IN)
+                return butler.get(ref)
+
+        def _put(self, value, ref):
+            self._checkMembership(ref, RefDirection.OUT)
+            butler.put(value, ref)
+
+        self._get = types.MethodType(_get, self)
+        self._put = types.MethodType(_put, self)
+
+    def get(self, dataset: typing.Union[InputQuantizedConnection, typing.List[DatasetRef], DatasetRef]):
+        """Fetches data from the butler
+
+        Parameters
+        ----------
+        dataset : `InputQuantizedConnection` or `list` of `lsst.daf.butler.DatasetRef`
+                or `lsst.daf.butler.DatasetRef`
+            This argument may either be an `InputQuantizedConnection` which describes
+            all the inputs of a quantum, a list of `lsst.daf.butler.DatasetRef`, or
+            a single `lsst.daf.butler.DatasetRef`. The function will get and return
+            the corresponding datasets from the butler.
+
+        Raises
+        ------
+        ValueError
+            If a `DatasetRef` is passed to get that is not defined in the quantum object
+        """
+        if isinstance(dataset, InputQuantizedConnection):
+            retVal = {}
+            for name, ref in dataset:
+                if isinstance(ref, list):
+                    val = []
+                    for r in ref:
+                        val.append(self._get(r))
+                else:
+                    val = self._get(ref)
+                retVal[name] = val
+            return retVal
+        elif isinstance(dataset, list):
+            return [self._get(x) for x in dataset]
+        elif isinstance(dataset, DatasetRef) or isinstance(dataset, DeferredDatasetRef):
+            return self._get(dataset)
+        else:
+            raise TypeError("Dataset argument is not a type that can be used to get")
```

Comment:
I'm not a big fan of the `isinstance`-based switch, it feels a bit fragile for my taste. But I'm not sure anything can be done to avoid it?

Reply:
In this case I dont think so if we are going to try and preserve this sort of overload. That is until we get to python 3.8. If (when?) we can take that as the base line, there is a nice built in single dispatch for methods. That will make the code cleaner and more extensible, but fundamentally we want this method overload based on the type of the input.